Event log

krystian-hebel opened the pull request:
This adds support for the DRTM TPM event log, both for TPM1.2 (aka SHA1 Event
Log Entry Format [1]) and TPM2.0 (aka Crypto Agile Log Entry Format [2]).

It is up to the bootloader to provide address and size of the event log. DRTM
specification [3] defines a new ACPI table containing this information, as well
as other pointers and data related to DRTM. Bootloader may prepare and pass
another block of memory if this table is not present.

If a bootloader does not fill up the event log address or size, LZ will not log
any entries, but it will continue to extend PCRs nonetheless.

[1] https://www.trustedcomputinggroup.org/wp-content/uploads/TCG_PCClientImplementation_1-21_1_00.pdf
[2] https://trustedcomputinggroup.org/wp-content/uploads/EFI-Protocol-Specification-rev13-160330final.pdf
[3] https://trustedcomputinggroup.org/wp-content/uploads/TCG_D-RTM_Architecture_v1-0_Published_06172013.pdf

krystian-hebel committed:
event_log: add fields for hash of LZ to the lz_header

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
event_log: add code for initializing and filling the DRTM TPM event log

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
main: log PCR extend operations in DRTM TPM event log

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

dpsmith commented on head.S:
```diff
@@ -256,13 +256,22 @@ ENDDATA(l4_identmap)
+	.long	0			/* Event Log size */
+	.fill	0x14		/* MSB Key Hash */
+
+	/* Hashes of LZ to be written to DRTM TPM event log */
```
Why not have a field for algo id and then reserve space for sha512 (largest hash)?

krystian-hebel replied:
This right now has the exact layout as is used by new (agile) log format, so it can be written with simple `memcpy`, while old (SHA1) format takes just one of these. We need both because we don't know the TPM family in advance.

krystian-hebel replied:
After the discussion on Wire I decided to change it with new bootloader data structure in one of the following PRs, for now I'd rather leave it as it is to have some base to work on

dpsmith commented on event_log.c:
```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
```
Is this a copy of the log code from Secure Launch? If you made modifications, you should add yourself at least as an author.

krystian-hebel replied:
Done

dpsmith commented on early_tpm.c:
```diff
@@ -98,6 +54,8 @@ u8 *tpmb_reserve(struct tpmbuff *b)
 
 void tpmb_free(struct tpmbuff *b)
 {
+	memset(b->head, 0, b->len);
```
LGTM is complaining that this is an implicit declaration.

krystian-hebel replied:
Yes, I know, this is also for the previous PR so I'll need to fix it there and rebase.

krystian-hebel replied:
Done

krystian-hebel committed:
event_log.c: make the log format compatible with TXT

In order to simplify adding new entries and to minimize the number of
differences between AMD and Intel, the event logs have actually two headers,
both for TPM 1.2 and 2.0.

For TPM 1.2 this is TCG_PCClientSpecIDEventStruct [1] with Intel's own
TXT-specific header embedded inside its 'vendorInfo' field. The offset to
this field is added to the base address in AMD path in kernel, making the
code for adding new events the same for both vendors.

TPM 2.0 in TXT uses HEAP_EVENT_LOG_POINTER_ELEMENT2_1 structure, which is
normally constructed on the TXT stack [2]. For AMD, this structure is put
inside TCG_EfiSpecIdEvent [3], also in 'vendorInfo' field. The actual offset
to this field depends on number of hash algorithms supported by the event
log.

[1] https://www.trustedcomputinggroup.org/wp-content/uploads/TCG_PCClientImplementation_1-21_1_00.pdf
[2] http://www.intel.com/content/dam/www/public/us/en/documents/guides/intel-txt-software-development-guide.pdf
[3] https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClientSpecPlat_TPM_2p0_1p04_pub.pdf

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

rossphilipson commented on main.c:
```diff
@@ -139,25 +140,28 @@ static void print_p(const void * unused) { }
-		print("PCR extended\n");
+		tpm_extend_pcr(tpm, pcr, TPM_HASH_ALG_SHA256, &sha256_hash[0]);
+
+		log_event_tpm20(pcr, hash, sha256_hash, ev);
```
It looks like SHA1 is extended for both 1.2 and 2.0 unconditionally. The event sent to the event log should also contain both hashes to be an accurate record of what went into the TPM. This is also consistent with the way the ACM and the SL code in the kernel currently handle things.

krystian-hebel replied:
And that is what this function does: `hash` is SHA1 and `sha256_hash` is SHA256. As both algorithms are required by TPM spec there is no reason to not do so, as SHA1 functions are compiled in anyway.

rossphilipson replied:
Oh I see. Sorry I missed that. Perfect, thanks.

dpsmith commented:
This was merged under PR #65
